core/src/main/scala/caliban/schema/Step.scala

```diff
@@ -77,8 +77,8 @@ object Step {
       }
 }
 
-sealed trait ReducedStep[-R] { self =>
-  def isPure: Boolean = false
+sealed abstract class ReducedStep[-R] { self =>
+  final def isPure: Boolean = self.isInstanceOf[PureStep]
```

Comment:
Could we make it a Val?

Reply:
I thought about it, although in most cases, we call the `isPure` method on a subset of all the `ReducedStep`s (e.g., in a list we only call it on the first element). So the performance ends up slightly worse if we make it a `val`